### PR TITLE
Added a Monte Carlo simulation

### PR DESCRIPTION
**Pull request by MatteoRaso**
### **Describe your change:**

Uses a Monte Carlo simulation to create an estimate of pi.

* [X] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [X] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [X] All new Python files are placed inside an existing directory.
* [X] All filenames are in all lowercase characters with no spaces or dashes.
* [X] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: maths/montecarlo.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    circle_dots = 0
+    
+    #A local function to see if a dot lands in the circle.
+    def circle(x: float, y: float): 
+        distance_from_centre = np.sqrt((x ** 2) + (y ** 2))
+         #Our circle has a radius of 1, so a distance greater than 1 would land outside the circle.
+        if distance_from_centre > 1:
+            return False
+        else:
+            return True
```
Comment (cclauss) [inline maths/montecarlo.py]: Replace 4 lines with one...  `return distance_from_centre <= 1`

File: maths/montecarlo.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            
+    for i in range(0, iterations):
+        #Random float from -1.0 to 1.0.
+        x = np.random.uniform(-1.0, 1.0) 
+        y = np.random.uniform(-1.0, 1.0)
+        if circle(x,y) == True:
+            circle_dots += 1
```
Comment (cclauss) [inline maths/montecarlo.py]:
```python
from numpy.random import uniform
circle_dots = sum(int(circle(uniform(-1.0, 1.0), uniform(-1.0, 1.0))) for _ in range(iterations))
```

File: maths/montecarlo.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    5. Multiply this value by 4 to get your estimate of pi.
+    6. Print the estimated and numpy value of pi
+    """
+    import numpy as np
+    circle_dots = 0
+    
+    #A local function to see if a dot lands in the circle.
+    def circle(x: float, y: float): 
+        distance_from_centre = np.sqrt((x ** 2) + (y ** 2))
+         #Our circle has a radius of 1, so a distance greater than 1 would land outside the circle.
```
Comment (cclauss) [inline maths/montecarlo.py]: Lines should be wrapped at 88 characters.

File: maths/montecarlo.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    3. For each iteration, place a dot anywhere in the square.
+    3.1 Record the number of dots within the circle.
+    4. After all the dots are placed, divide the dots in the circle by the total.
+    5. Multiply this value by 4 to get your estimate of pi.
+    6. Print the estimated and numpy value of pi
+    """
+    import numpy as np
+    circle_dots = 0
+    
+    #A local function to see if a dot lands in the circle.
```
Comment (cclauss) [inline maths/montecarlo.py]: Please format the code with psi/ black as discussed in CONTRIBUTING.md.

**Comment by MatteoRaso** [comment]
I have made the requested changes to the program.

File: maths/montecarlo.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+"""
+@author: MatteoRaso
+"""
+
+
+def piEstimator(iterations: int):
```
Comment (onlinejudge95) [inline maths/montecarlo.py]: Follow PEP8 naming conventions

File: maths/montecarlo.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    """An implementation of the Monte Carlo method used to find pi.
+    1. Draw a 2x2 square centred at (0,0).
+    2. Inscribe a circle within the square.
+    3. For each iteration, place a dot anywhere in the square.
+    3.1 Record the number of dots within the circle.
+    4. After all the dots are placed, divide the dots in the circle by the total.
+    5. Multiply this value by 4 to get your estimate of pi.
+    6. Print the estimated and numpy value of pi
+    """
+    from numpy import pi, sqrt
```
Comment (onlinejudge95) [inline maths/montecarlo.py]: How about keeping them in the module's global namespace?

File: maths/montecarlo.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    from random import uniform
+
+    circle_dots = 0
+
+    # A local function to see if a dot lands in the circle.
+    def circle(x: float, y: float):
```
Comment (onlinejudge95) [inline maths/montecarlo.py]: Why nesting function, if not using them as decorators or closure?

Comment (cclauss) [reply maths/montecarlo.py]: It is legit to only allow the nested function to be called from it’s parent function.

Comment (onlinejudge95) [reply maths/montecarlo.py]: But what does it achieves, maybe define the nested function on the same level as parent function, or maybe using a lambda(at least in this case).

Comment (cclauss) [reply maths/montecarlo.py]: It is only designed to be called inside the main function.  I find a nested function to be more natural and understandable than a lambda.

**Comment by MatteoRaso** [comment]
> I am uncomfortable with an algorithm that provides variable precision but I guess that is the nature of this approach.

Sometimes that's an unavoidable tradeoff in programming. It is still quite reliable despite the precision being variable, however, since the large sample size makes it extremely unlikely for the estimate to deviate significantly far from the true value.

**Comment by cclauss** [comment]
I approve.

**Comment by cclauss** [comment]
https://realpython.com/inner-functions-what-are-they-good-for

**Pull request merged**